lib/sandbox.jail.ml

```diff
@@ -0,0 +1,157 @@
+open Lwt.Infix
+open Sexplib.Conv
+
+let ( / ) = Filename.concat
+
+type t = {
+  lock : Lwt_mutex.t;
+}
+
+type config = {
+  dummy: bool;
+} [@@deriving sexp]
+
+(* Find out the user name to use within the jail, by parsing the
+   /etc/passwd file within the jail filesystem. This is roughly
+   equivalent to what Unix.getpwuid would do.
+   Note that the gid is currently ignored. *)
+let jail_username rootdir config =
+  match config.Config.user with
+  | `Windows w -> w.name
+  | `Unix { uid; _ } ->
+    let pwdfile = rootdir / "etc" / "passwd" in
+    let uidstr = string_of_int uid in
+    let rec parse_line ch =
+      match In_channel.input_line ch with
+      | None -> None
+      | Some line ->
+        let fields = String.split_on_char ':' line in begin
+          match fields with
+          | name :: _pass :: uid :: _ ->
+            if uid = uidstr then Some name else parse_line ch
+          | _ -> parse_line ch
+      end
+    in
+    match In_channel.with_open_text pwdfile parse_line with
+    | None -> Fmt.failwith "No user found for uid %d" uid
+    | Some name -> name
+
+(* Compute the complete set of arguments passed to the jail(8) command:
+   jail username, jail path, command to run, etc. *)
+let jail_options config rootdir =
+  let username = jail_username rootdir config in
+  let commandline =
+    let env = List.rev_map (fun (k, v) -> k ^ "=" ^ v) config.env in
+    let commandline =
+      match config.argv with
+      | _shell :: "-c" :: args ->
+        (* This overrides the default shell set in build.ml, since bash is
+           not installed by default (and would be in /usr/local/bin anyway) *)
+       [ "/bin/sh" ; "-c" ] @ List.rev (List.rev_map Filename.quote args)
+      | args -> List.rev (List.rev_map Filename.quote args)
```

Comment:
If you're doing `List.rev (List.rev_map f xs)`, maybe you just want `List.map` instead?

Reply:
`List.map` is not tail-recursive, and there is no guarantee the list won't be quite large.